From: Piotr Nosek <piotr.nosek@erlang-solutions.com>
Date: Tue, 1 Sep 2015 12:12:32 +0200
Subject: [PATCH] Fixes for OTP 18

Remove warnings_as_errors

Use erlang:system_time/1 for OTP 18.x

diff --git a/.travis.yml b/.travis.yml
index 65f49b9..20d00f2 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -6,8 +6,6 @@ notifications:
   webhooks: http://basho-engbot.herokuapp.com/travis?key=8f07584549e458d4c83728f3397ecbd4368e60a8
   email: eng@basho.com
 otp_release:
-  - R16B02
-  - R16B01
-  - R15B03
-  - R15B02
-  - R15B01
+  - 18.1
+  - 17.5
+  - R16B03
diff --git a/rebar.config b/rebar.config
index 1a5e7ed..8e917bc 100644
--- a/rebar.config
+++ b/rebar.config
@@ -1,14 +1,26 @@
-{require_otp_vsn, "R15|R16|17"}.
+{require_otp_vsn, "R15|R16|17|18"}.
 
 {cover_enabled, true}.
+
 {eunit_opts, [verbose]}.
-{erl_opts, [warnings_as_errors, debug_info, nowarn_deprecated_type]}.
+
+{erl_opts, [
+    debug_info,
+    warnings_as_errors,
+    {platform_define, "^[0-9]+", namespaced_types},
+    {platform_define, "^18", deprecated_now}
+]}.
 {deps, [
-        {riak_pb, "2.1.0.2", {git, "git://github.com/basho/riak_pb", {tag, "2.1.0.2"}}}
-       ]}.
-{edoc_opts, [{stylesheet_file, "priv/edoc.css"},
-             {preprocess, true}]}.
+    {riak_pb, "2.1.0.7", {git, "git://github.com/basho/riak_pb", {tag, "2.1.0.7"}}}
+]}.
+
+{edoc_opts, [
+    {stylesheet_file, "priv/edoc.css"},
+    {preprocess, true}
+]}.
+
 {clean_files, ["doc/*.html", "doc/*.png", "doc/edoc-info", "doc/*.css"]}.
+
 {xref_checks, [undefined_function_calls, undefined_functions,
                deprecated_function_calls, deprecated_functions,
                locals_not_used]}.
diff --git a/src/riakc_pb_socket.erl b/src/riakc_pb_socket.erl
index 2772b4e..59aca82 100644
--- a/src/riakc_pb_socket.erl
+++ b/src/riakc_pb_socket.erl
@@ -130,6 +130,18 @@
 -record(request, {ref :: reference(), msg :: rpb_req(), from, ctx :: ctx(), timeout :: timeout(),
                   tref :: reference() | undefined }).
 
+-ifdef(namespaced_types).
+-type request_queue_t() :: queue:queue(#request{}).
+-else.
+-type request_queue_t() :: queue().
+-endif.
+
+-ifdef(deprecated_now).
+-define(NOW, erlang:system_time(micro_seconds)).
+-else.
+-define(NOW, erlang:now()).
+-endif.
+
 -type portnum() :: non_neg_integer(). %% The TCP port number of the Riak node's Protocol Buffers interface
 -type address() :: string() | atom() | inet:ip_address(). %% The TCP/IP host name or address of the Riak node
 -record(state, {address :: address(),    % address to connect to
@@ -141,7 +153,7 @@
                 keepalive = false :: boolean(), % if true, enabled TCP keepalive for the socket
                 transport = gen_tcp :: 'gen_tcp' | 'ssl',
                 active :: #request{} | undefined,     % active request
-                queue :: queue() | undefined,      % queue of pending requests
+                queue :: request_queue_t() | undefined,      % queue of pending requests
                 connects=0 :: non_neg_integer(), % number of successful connects
                 failed=[] :: [connection_failure()],  % breakdown of failed connects
                 connect_timeout=infinity :: timeout(), % timeout of TCP connection
@@ -2186,7 +2198,7 @@ remove_queued_request(Ref, State) ->
     end.
 
 %% @private
-mk_reqid() -> erlang:phash2(erlang:now()). % only has to be unique per-pid
+mk_reqid() -> erlang:phash2(?NOW). % only has to be unique per-pid
 
 %% @private
 wait_for_list(ReqId) ->
